### DISCUSSION TITLE
Beginner trying Remix. Stuck on getting Mapbox working. Anyone got Mapbox working with Remix?

### DISCUSSION THREAD

**Discussion by danielcolinjames** [remix-run/remix#1991] [Q&A] [answered]
Hi there! I'm currently attempting to see if Remix could be a suitable framework for the [Helium Explorer](https://explorer.helium.com), and haven't been able to get nearly anything running when trying to port [the existing code](https://github.com/helium/explorer) over. It's currently using Next.js, with React-Router and a custom server.js (because we wanted to have nested routes, persistent layouts, etc., and a custom server because something was breaking on Heroku without that.)

The nested routes support of Remix is very appealing, since that's how we've built the Explorer in terms of UX—persistent parts of the UI, each sub-part tied to a new URL route. We've had to break out of Next's default routing and use more of a SPA structure to achieve that.

The thing that's led me to trying Remix is that we want to get meta tags working properly with the Explorer, and it's quite difficult and complicated to get working right now with our current project structure.

In my first attempts to get the codebase ported over to Remix, I was getting all sorts of weird and un-googleable errors when starting from an empty Remix project and trying to add this react-mapbox library: https://github.com/alex3165/react-mapbox-gl, which is the one we are currently using on Explorer.

So I tried using this library instead: https://github.com/visgl/react-map-gl since it appears to be better maintained and more frequently updated.

But I still can't seem to be able to get a map showing up no matter what I try. First I was getting a bunch of this error: https://github.com/remix-run/remix/issues/1618.

Now I'm just getting a page with a seemingly empty div where the map should be, and an error in the console about how I need to provide the Map an API token (which I am doing). I tried doing it with env vars, but Remix's approach to env vars somewhat baffles me, and I kept getting either "`process` is undefined", or "`window` is undefined" errors. Manually adding the token as a prop to the Map component also didn't work somehow.

And I'm also still a bit confused about how routes in Remix work, and particularly the difference between what should be in `app/routes/whatever.jsx` vs `app/routes/whatever/index.jsx` makes my head hurt. In Next, between `pages/whatever.jsx` and `pages/whatever/index.jsx` it's one or the other, unless I'm mistaken. In Remix, it seems to be both, but they do different things?

Anyway, I think I'm at the end of my rope trying to get Remix working for now. I think maybe Create React App is better suited for the Helium Explorer right now, but would love any comments that can point me in the right direction for any of the issues I'm facing.

I figured I should post my experience so far in case anyone else finds this from a Google search who is attempting to do what I'm attempting to do. I think map-based web apps is a great use case for Remix. Or in case I'm missing some simple things and just need to push a little more through confusing-error-land.

Thanks!

**Comment by kiliman** [comment, marked as answer]
I created an example app that uses Mapbox.

https://github.com/kiliman/remix-mapbox

You can see it live here: https://remix-mapbox.herokuapp.com/

BTW: Remix v1.2.3 will automatically load from .env file if present. You still need to set your environment variables in production.

---

As for your routes question:

If you have a file `app/routes/whatever.jsx`, then Remix will serve up that route when you to go `/whatever`. That's how Remix and Next.js works. However, Remix also supports *nested* layouts. This means, that you can have a layout and inside that layout, render a child route.

So in your example, if you create a folder with the *same name* as the layout file
```
routes/whatever.tsx  <- layout file (render child in <Outlet/>)
routes/whatever/index.tsx  <- child route (content will render in parent <Outlet/>)
routes/whatever/another.tsx  <- another child route (content will render in parent <Outlet/>)
```
```ts
// routes/whatever.tsx
export default function Layout() {
  return (
    <div>
      <h1>Whatever Layout</h1>
      <Outlet/>
    </div>
  )
}
// routes/whatever/index.tsx
export default function Index() {
  return <h2>Index</h2>  // this will be rendered in parent <Outlet/>
}
```
So when you go to `/whatever`, Remix will render the layout `whatever.tsx` and the child route `whatever/index.tsx` which will look like this:
```html
<div>
  <h1>Whatever Layout</h1>
  <h2>Index</h2>
<div>
```

BTW: Sorry that your first experience with Remix was a rough one.
